Build_Process/LinkerScripts.md

```diff
@@ -0,0 +1,196 @@
+- data, with both read and write permissions.
+
+*But what about the bss, and other zero-initialized data?* Well that's stored in the data program header. Program headers have two variables for their size, one is the file size and the other is their memory size. If the memory size is bigger than the file size, that memory is expected to be zeroed (as per the elf spec), and thus the bss can be placed there!
+
+### Example
+An example of what a program headers section might look like:
+
+```
+/* This declares our program headers */
+PHDRS
+{
+    text   PT_LOAD;
+    rodata PT_LOAD;
+    data   PT_LOAD;
+}
+```
+
+This example is actually missing the flags field that sets the permissions, but modern linkers will see these common names like `text` and `rodata` and give them default permissions.
+
+Of course, you can (and in best practice, should) set them manually:
+
+```
+PHDRS
+{
+    /* flags bits: 0 = execute, 1 = write, 2 = read */
+    text   PT_LOAD FLAGS((1 << 0) | (1 << 2));
+    rodata PT_LOAD FLAGS((1 << 2));
+    data   PT_LOAD FLAGS((1 << 1) | (1 << 2));
+}
+```
+
+## Sections
+While program headers are a fairly course mechanism for telling the program loader what it needs to do in order to get our program running, sections allow us a lot of control over how the code and data within those areas is arranged.
+
+### The '.' Operator
+When working with sections, we'll want to control where sections are placed in memory. We can use absolute addresses, however this means we'll need to update the linker script manually everytime things change in size. Enter the dot operator (`.`), it represents the current VMA (remember this is the runtime address). 
+
+The linker will automatically increment this when placing sections, and the current vma is read/write, so we have complete control over this process if we want.
+
+Often simply setting it at the beginning (before the first section) is enough, like so:
+
+```
+SECTIONS
+{
+    . = 0xFFFFFFFF80000000;
+    /* section descritions go here */
+}
+```
+
+If you're wondering why most higher half kernels are loaded at this address, it's because it's the upper-most 2GB of the 66-bit address space
+
+### Incoming vs Outgoing Sections
+A section description has 2 parts: incoming sections (all your object files), and how they are placed into outgoing sections (the final output file).
+
+Let's consider the following example:
+
+```
+.text :
+{
+    *(.text*)
+    /* ... more input sections ... */
+} :text
+```
+
+First we give this output section a name, in this case it's `.text`. To the right of the colon is where we would place any extra attributes for this section or override the LMA. By default the LMA will be the same as the VMA, but if you need you can override it here, with something like `.text : AT(0x1234)`. The text section would now have the VMA of 0 unless we've overriden it like before, but LMA of 0x1234. Unless you know you need to change the LMA, let this part empty like the example.
```

Comment:
`AT()` is used for overriding the LMA, correct :)
Linker scripts are focused around changing the VMA, since that's what most programs will be dealing with, so changing the vma is simple: you use the dot operator!

```
. = 0x1234;
```
This would set the current VMA, and the vma will also be incremented by the size of any sections. For example if you added a section after that line above that was 0x1000 bytes, the new VMA would be 0x2234. If you used ALIGN(4K) on the section that followed, it would start at VMA 0x3000.

I can make this clearer in the text.

Reply:
yeah please! :)